src/reference/asciidoc/testing.adoc

```diff
@@ -2,12 +2,278 @@
+    assertThat(result, is(notNullValue()));
+    assertThat(result, hasPayload(is(instanceOf(byte[].class))));
+    assertThat(result, hasPayload(SAMPLE_CONTENT.getBytes(DEFAULT_ENCODING)));
+}
+----
+
+The `MockitoMessageMatchers` factory can be used for mocks stubbing and verifications:
+
+[source,java]
+----
+static final Date SOME_PAYLOAD = new Date();
+
+static final String SOME_HEADER_VALUE = "bar";
+
+static final String SOME_HEADER_KEY = "test.foo";
+...
+Message<?> message = MessageBuilder.withPayload(SOME_PAYLOAD)
+                .setHeader(SOME_HEADER_KEY, SOME_HEADER_VALUE)
+                .build();
+MessageHandler handler = mock(MessageHandler.class);
+handler.handleMessage(message);
+verify(handler).handleMessage(messageWithPayload(SOME_PAYLOAD));
+verify(handler).handleMessage(messageWithPayload(is(instanceOf(Date.class))));
+...
+MessageChannel channel = mock(MessageChannel.class);
+when(channel.send(messageWithHeaderEntry(SOME_HEADER_KEY, is(instanceOf(Short.class)))))
+        .thenReturn(true);
+assertThat(channel.send(message), is(false));
+----
+
+More utils can be added or migrated eventually.
+For example `RemoteFileTestSupport` implementations for the (S)FTP tests can be moved from the `test` directory of the particular modules to this publicly available `spring-integration-test-support` artifact.
+
+[[test-context]]
+=== Spring Integration and test context
+
+Typically tests for Spring applications use the Spring Test Framework and since Spring Integration is fully based on the Spring Framework foundation, everything what we can do with the Spring Test Framework is applied as well when we come to the task for testing integration flows.
+The `org.springframework.integration.test.context` package provides some components for enhancing test context for integration needs.
+First of all we should configure our test class with particular `@SpringIntegrationTest` annotation to enable Spring Integration Test Framework:
+
+[source,java]
+----
+@RunWith(SpringRunner.class)
+@SpringIntegrationTest(noAutoStartup = {"inboundChannelAdapter", "*Source*"})
+public class MyIntegrationTests {
+
+    @Autowired
+    private MockIntegrationContext mockIntegrationContext;
+
+}
+----
+
+The `@SpringIntegrationTest` populates `MockIntegrationContext` bean which can be autowired to the test class for further usage.
+With the provided `noAutoStartup` option, the Spring Integration Test Framework switches off endpoints, which are matched to the provided patterns, from the auto-startup during application context refreshing.
+This is useful, when we would like do not have real connections to the target systems from Inbound Channel Adapters, for example AMQP Inbound Gateway, JDBC Polling Channel Adapter, WebSocket Message Producer in client mode etc.
+
+The `MockIntegrationContext` is aimed to be used in the target test-cases for modifications beans in the real application context, for example those stopped endpoints can be replaced with mocks:
```

Comment:
Yes, that's true. An endpoint will act as it is with regular application. But no one endpoint moves to that beans map, unless an explicit `instead()` call. If you don't want to have endpoint started after replacement, use an overloaded `instead` version. Otherwise the replaced part in that endpoint will start act immediately after replacement. But yeah, I think I'll move replacement to the very end of `instead` to avoid some race conditions.